ISPN-5924 and ISPN-5925 GlobalStateManager and PersistentUUID

tristantarrant opened the pull request:
https://issues.jboss.org/browse/ISPN-5924
https://issues.jboss.org/browse/ISPN-5925

Before I come along with the actual graceful cache shutdown / start, here are two building blocks.

The GlobalStateManager sits in the GCR lifecycle and manages global and per-cache state persistence.

The LocalUUID is the unique node identifier which survives restarts.

danberindei commented:
@tristantarrant Lots of failures caused by this NPE:

    Caused by: java.lang.NullPointerException
    at org.infinispan.globalstate.impl.GlobalStateManagerImpl.start(GlobalStateManagerImpl.java:71)

danberindei commented on core/src/main/java/org/infinispan/globalstate/impl/GlobalStateManagerImpl.java:
```diff
@@ -0,0 +1,132 @@
+      File stateFile = getStateFile(state.getScope());
+      try (PrintWriter w = new PrintWriter(stateFile)) {
+         state.forEach((key, value) -> {
+            w.printf("%s=%s%n", key, value);
```
Does this mean a state value must not contain '\n', and a key must not contain '='? Perhaps the `ScopedPersistentState` javadoc should explicitly say what the allowed characters are?

tristantarrant replied:
I'd rather encode them.

danberindei commented on core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java:
```diff
@@ -641,6 +652,21 @@ private int getGlobalTimeout() {
+
+   @Override
+   public void prepareForShutdown(ScopedPersistentState state) {
+      state.setProperty("uuid", localUUID.toString());
```
Shouldn't we write this immediately after startup, in case the node crashes?

tristantarrant replied:
Yes, that's what I have in my second set of patches, but it makes more sense to do it here.

danberindei commented on core/src/main/java/org/infinispan/topology/LocalUUID.java:
```diff
@@ -0,0 +1,76 @@
+import org.infinispan.remoting.transport.Address;
+
+/**
+ * LocalUUID. A special {@link Address} UUID whose purpose is to remain unchanged across node
```
I don't think you need to say the class name in the javadoc :)

TBH `Local` doesn't seem relevant either, perhaps `PersistentUUID` would sound better?

tristantarrant replied:
PersistentUUID is much better.

tristantarrant commented:
Done @danberindei 
* I have renamed LocalUUID to PersistentUUID
* I have added unicode escape/unescape to the state persister
* State is persisted immediately after the LTM has chosen a PersistentUUID

danberindei commented on core/src/main/java/org/infinispan/globalstate/impl/GlobalStateManagerImpl.java:
```diff
@@ -0,0 +1,136 @@
+      File stateFile = getStateFile(state.getScope());
+      try (PrintWriter w = new PrintWriter(stateFile)) {
+         state.forEach((key, value) -> {
+            w.printf("%s=%s%n", key, Util.unicodeEscapeString(value));
```
I think the key should be escaped as well.

danberindei commented on core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsAddressCache.java:
```diff
@@ -34,11 +37,22 @@
+      persistentUUIDCache.put(jgAddress, localUUID);
+   }
+
+   public static PersistentUUID getLocalUUID(Object address) {
```
s/LocalUUID/PersistentUUID/

tristantarrant commented:
@danberindei :
* escaped key
* added a commit to fix some leftover renames from ISPN-5851
* removed traces of localuuid :)

wburns commented on commons/src/main/java/org/infinispan/commons/util/Util.java:
```diff
@@ -699,6 +699,10 @@ private static void addHexByte(StringBuilder buf, byte b) {
    }
 
+   private static void addSingleHexByte(StringBuilder buf, byte b) {
+      buf.append(HEX_VALUES.charAt(b));
```
Guessing we may want to & with 0x0F to be safe?

tristantarrant commented:
Fixed @wburns

wburns commented:
Seems fine to me, will pull just after the previous CI run completes.

wburns commented:
Pulling...

wburns commented:
Integrated into master, thanks !
